### DISCUSSION TITLE
Blend image-label to check the segmentation region

### DISCUSSION THREAD

**Discussion by HripsimeS** [Project-MONAI/tutorials#621] [Q&A] [answered]
Dear MONAI's team @wyli @Nic-Ma @yiheng-wang-nv 

I am trying to optimize your method to blend the image and label to check the segmentation region (cells 11-12).
https://github.com/Project-MONAI/tutorials/blob/master/modules/transform_visualization.ipynb

Here is my notebook, please take a look https://github.com/HripsimeS/Computer-Vision/blob/main/3DUnet_GitHub..ipynb

It worked with image and true mask (cells 10-11) as they had the same shape (single channel), although on the 3rd column blend of all labels have the same red color (I have 4 labels excluding the background). 
But when I am trying to do blend of image and predicted mask (cells 30-31), I get an error because my predicted mask has 5 channels and image has a single channel. Can you please help to figure out how to fix this channel issue and how to have different colors of each labels on blend images for the 3rd column. Look forward to hearing from you!

**Comment by rijobro** [comment, marked as answer]
Strange. Can paste the results of `check_data["mask"][0].dtype` and `torch.unique(check_data["mask"][0])`? Also `check_data["image"].max()`

**Reply by HripsimeS** [reply]
@rijobro thanks for the advice concerning the colors, I changed rescale_array=True and each label I can see with different color 🥇 
But I am still struggling to blend the image with predicted mask due to channels issue, 30-31 cells in my notebook.

rety = blend_images(image=original_image, label=test_outputs[0], alpha=0.5, cmap="hsv", rescale_arrays=True)

for i in range(5, 10):
    
    slice_index = 20 * i
    plt.figure("blend image and mask", (12, 4))
    plt.subplot(1, 3, 1)
    plt.title(f"image slice {slice_index}")
    plt.imshow(original_image[:, :, slice_index], cmap="gray")
    plt.subplot(1, 3, 2)
    plt.title(f"predicted mask slice {slice_index}")
    plt.imshow(torch.argmax(test_outputs[0], dim=0).detach().cpu()[:, :, slice_index])
    plt.subplot(1, 3, 3)
    plt.title(f"blend slice {slice_index}")
    # switch the channel dim to the last dim
    plt.imshow(torch.moveaxis(rety[:, :, :, slice_index], 0, -1))
    plt.show()

**Reply by rijobro** [reply]
The blend_images function expects 1 channel as input. So you want to do [0] to remove the batch dimension (you're already doing this) and then `.argmax(0)` to convert your one-hotted results to a single channel. You'll probably need the `keep_dim=True` argument, too.

**Reply by HripsimeS** [reply]
@rijobro I have tried your suggestion test_outputs[0].argmax(0), but still got an error "Label should have 1 channel". Where I should add keep_dim=True?  Please can you copy my code above and resend me it again with your modifications you think could be helpful :)

**Reply by rijobro** [reply]
```python
test_output_argmax = torch.argmax(test_outputs[0], dim=0, keepdim=True)
print(test_output_argmax.shape)  # should be [C,H,W,D] where C==1

rety = blend_images(image=original_image, label=test_output_argmax, alpha=0.5, cmap="hsv", rescale_arrays=True)
```

**Reply by HripsimeS** [reply]
@rijobro thanks a lot, now my label shape is torch.Size([1, 512, 512, 313]). But now got other error "Image should have 1 or 3 channels" because image shape is (512, 512, 313). 
I assume the error comes from this code original_image = loader(test_data[0]["image_meta_dict"]["filename_or_obj"])[0]
When I remove [0] near test_data, I receive "list indices must be integers or slices, not str"
When I remove [0] from the end, I receive the following "'tuple' object has no attribute 'shap"

**Reply by rijobro** [reply]
Add channel component to your original image:
```python
rety = blend_images(image=original_image[None], label=test_output_argmax, alpha=0.5, cmap="hsv", rescale_arrays=True)
```

**Reply by rijobro** [reply]
You mix `[B,C,H,W,D]`, `[C,H,W,D]` and `[H,W,D]` throughout your code. It would be best to be clear about the shape required at each step to avoid these problems. Also there's no need to re-load your image (`original_image = loader(test_data[0]["image_meta_dict"]["filename_or_obj"])[0]`) as it is already `image_data["image"]` in a different shape.

**Reply by HripsimeS** [reply]
@rijobro thanks a lot, eventually it worked out 👍 🥇 


![done](https://user-images.githubusercontent.com/42246765/159728374-3951ab24-0530-4b98-903c-3c5854bc2f95.png)

**Reply by rijobro** [reply]
Looks great!
